<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module meta</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong>meta</strong></big></big> (version 1.0.1)</font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="file:/media/jean-noel/98257acf-f310-4578-90bb-4121919162c6/Projet_Anne/strali/meta.py">/media/jean-noel/98257acf-f310-4578-90bb-4121919162c6/Projet_Anne/strali/meta.py</a></font></td></tr></table>
    <p><tt>Meta&nbsp;module&nbsp;handles&nbsp;the&nbsp;data&nbsp;from&nbsp;a&nbsp;pool&nbsp;of&nbsp;genes.</tt></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#aa55cc">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Modules</strong></big></font></td></tr>
    
<tr><td bgcolor="#aa55cc"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><table width="100%" summary="list"><tr><td width="25%" valign=top><a href="os.html">os</a><br>
</td><td width="25%" valign=top></td><td width="25%" valign=top></td><td width="25%" valign=top></td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ee77aa">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Classes</strong></big></font></td></tr>
    
<tr><td bgcolor="#ee77aa"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl>
<dt><font face="helvetica, arial"><a href="builtins.html#object">builtins.object</a>
</font></dt><dd>
<dl>
<dt><font face="helvetica, arial"><a href="meta.html#MetaGene">MetaGene</a>
</font></dt></dl>
</dd>
</dl>
 <p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ffc8d8">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#000000" face="helvetica, arial"><a name="MetaGene">class <strong>MetaGene</strong></a>(<a href="builtins.html#object">builtins.object</a>)</font></td></tr>
    
<tr bgcolor="#ffc8d8"><td rowspan=2><tt>&nbsp;&nbsp;&nbsp;</tt></td>
<td colspan=2><tt>MetaGenes&nbsp;base&nbsp;class.<br>&nbsp;</tt></td></tr>
<tr><td>&nbsp;</td>
<td width="100%">Methods defined here:<br>
<dl><dt><a name="MetaGene-__init__"><strong>__init__</strong></a>(self, transcripts)</dt><dd><tt>Init&nbsp;with&nbsp;a&nbsp;list&nbsp;of&nbsp;transcripts.</tt></dd></dl>

<dl><dt><a name="MetaGene-__repr__"><strong>__repr__</strong></a>(self)</dt><dd><tt>Return&nbsp;repr(self).</tt></dd></dl>

<dl><dt><a name="MetaGene-allframes_counts"><strong>allframes_counts</strong></a>(self, bamfile, shift=12, side=5, only_total=False, title='All frames')</dt><dd><tt>Return&nbsp;the&nbsp;pooled&nbsp;counts&nbsp;in&nbsp;each&nbsp;frame&nbsp;of&nbsp;the&nbsp;CDSes<br>
of&nbsp;the&nbsp;transcripts.</tt></dd></dl>

<dl><dt><a name="MetaGene-regions_stats"><strong>regions_stats</strong></a>(self, bamfile, shift=12, side=5, title='Regions', only_total=False, colors=['black', 'gray', 'gray'])</dt><dd><tt>One&nbsp;loop&nbsp;over&nbsp;all&nbsp;transcripts&nbsp;to&nbsp;have&nbsp;the<br>
total&nbsp;number&nbsp;of&nbsp;reads&nbsp;on&nbsp;each&nbsp;region&nbsp;+&nbsp;the&nbsp;sizes.</tt></dd></dl>

<dl><dt><a name="MetaGene-start_stop_envs"><strong>start_stop_envs</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'], inner_interval=100, outer_interval=20, only_total=False)</dt><dd><tt>Return&nbsp;a&nbsp;dictionary&nbsp;with&nbsp;the&nbsp;start&nbsp;and&nbsp;stop&nbsp;meta&nbsp;coverage.</tt></dd></dl>

<dl><dt><a name="MetaGene-transcripts_expression"><strong>transcripts_expression</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'], sample_id=None, rnaseq=None, qpcr=None, title='Expression profile')</dt><dd><tt>compute&nbsp;the&nbsp;expression&nbsp;profile&nbsp;of&nbsp;the&nbsp;transcripts,&nbsp;normalising<br>
by&nbsp;rnaseq&nbsp;counts&nbsp;or&nbsp;qpcr&nbsp;abundances&nbsp;if&nbsp;provided.&nbsp;rnaseq&nbsp;/&nbsp;qpcr&nbsp;are<br>
exclusive.&nbsp;They&nbsp;must&nbsp;be&nbsp;pandas&nbsp;dataframes&nbsp;with&nbsp;genes&nbsp;id&nbsp;/&nbsp;name&nbsp;as<br>
index.</tt></dd></dl>

<hr>
Data descriptors defined here:<br>
<dl><dt><strong>__dict__</strong></dt>
<dd><tt>dictionary&nbsp;for&nbsp;instance&nbsp;variables&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
<dl><dt><strong>__weakref__</strong></dt>
<dd><tt>list&nbsp;of&nbsp;weak&nbsp;references&nbsp;to&nbsp;the&nbsp;object&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
</td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#55aa55">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Data</strong></big></font></td></tr>
    
<tr><td bgcolor="#55aa55"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><strong>APPRIS_LEVEL</strong> = 1<br>
<strong>APPRIS_P_ALL</strong> = 'appris_principal'<br>
<strong>APPRIS_P_LEV</strong> = 'appris_principal_1'<br>
<strong>ATTRIBUTES</strong> = 8<br>
<strong>BLIT</strong> = '<font color="#c040c0">\x1b</font>[3;34m'<br>
<strong>BLITUN</strong> = '<font color="#c040c0">\x1b</font>[4;34m'<br>
<strong>CDS</strong> = 'CDS'<br>
<strong>END</strong> = 4<br>
<strong>EXON</strong> = 'exon'<br>
<strong>FRAMES_CORREL</strong> = {0: {0: 0, 1: 1, 2: 2}, 1: {0: 2, 1: 0, 2: 1}, 2: {0: 1, 1: 2, 2: 0}}<br>
<strong>GENE</strong> = 'gene'<br>
<strong>GFF_COM</strong> = '#'<br>
<strong>GFF_SEP</strong> = '<font color="#c040c0">\t</font>'<br>
<strong>GRIT</strong> = '<font color="#c040c0">\x1b</font>[3;32m'<br>
<strong>ID_TAG</strong> = 'ID'<br>
<strong>INNER</strong> = 100<br>
<strong>KMERS</strong> = ['Total']<br>
<strong>MASK_BASE</strong> = 'N'<br>
<strong>NA</strong> = '<font color="#c040c0">\x1b</font>[0m'<br>
<strong>NAME_TAG</strong> = 'Name'<br>
<strong>N_0</strong> = 0<br>
<strong>ORDERED_REGIONS</strong> = ['UTR5', 'CDS', 'UTR3']<br>
<strong>ORIT</strong> = '<font color="#c040c0">\x1b</font>[3;33m'<br>
<strong>OUTER</strong> = 20<br>
<strong>PARENT_TAG</strong> = 'Parent'<br>
<strong>PHASE</strong> = 7<br>
<strong>RANK_TAG</strong> = 'exon_number'<br>
<strong>REIT</strong> = '<font color="#c040c0">\x1b</font>[3;31m'<br>
<strong>SCORE</strong> = 5<br>
<strong>SEQID</strong> = 0<br>
<strong>SHIFT</strong> = 12<br>
<strong>SIDE</strong> = 5<br>
<strong>SOURCE</strong> = 1<br>
<strong>START</strong> = 3<br>
<strong>STRAND</strong> = 6<br>
<strong>STRANDS_MAP</strong> = {'+': {'ali': 0, 'ali_rev': 16, 'one': 1}, '-': {'ali': 16, 'ali_rev': 0, 'one': -1}, '.': {'one': 1}}<br>
<strong>TAG_SEP</strong> = ';'<br>
<strong>TOTAL_KEYNAME</strong> = 'Total'<br>
<strong>TRANSCRIPT</strong> = 'transcript'<br>
<strong>TYPE</strong> = 2<br>
<strong>UTR</strong> = 'UTR'<br>
<strong>UTR3</strong> = 'UTR3'<br>
<strong>UTR5</strong> = 'UTR5'<br>
<strong>VAL_SEP</strong> = '='<br>
<strong>__email__</strong> = 'pierre.bertin@i2bc.paris-saclay.fr'<br>
<strong>__license__</strong> = 'GPL'<br>
<strong>__maintainer__</strong> = 'Pierre Bertin'<br>
<strong>__status__</strong> = 'Development'</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#7799ee">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Author</strong></big></font></td></tr>
    
<tr><td bgcolor="#7799ee"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%">Pierre&nbsp;Bertin</td></tr></table>
</body></html>