<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module noncoding_objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong>noncoding_objects</strong></big></big> (version 1.0.1)</font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="file:/media/jean-noel/98257acf-f310-4578-90bb-4121919162c6/Projet_Anne/strali/noncoding_objects.py">/media/jean-noel/98257acf-f310-4578-90bb-4121919162c6/Projet_Anne/strali/noncoding_objects.py</a></font></td></tr></table>
    <p><tt>A.&nbsp;Lopes&nbsp;project:&nbsp;The&nbsp;non&nbsp;coding&nbsp;regions&nbsp;associated&nbsp;objects.<br>
&nbsp;<br>
This&nbsp;module&nbsp;contains&nbsp;all&nbsp;objects&nbsp;relative&nbsp;to&nbsp;intergenic&nbsp;regions&nbsp;/&nbsp;ORFs,&nbsp;called<br>
IGR&nbsp;and&nbsp;IGORF&nbsp;respectively.</tt></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ee77aa">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Classes</strong></big></font></td></tr>
    
<tr><td bgcolor="#ee77aa"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl>
<dt><font face="helvetica, arial"><a href="biobjects.html#Feature">biobjects.Feature</a>(<a href="biobjects.html#StraGeRa">biobjects.StraGeRa</a>)
</font></dt><dd>
<dl>
<dt><font face="helvetica, arial"><a href="noncoding_objects.html#Igorf">Igorf</a>
</font></dt></dl>
</dd>
<dt><font face="helvetica, arial"><a href="builtins.html#object">builtins.object</a>
</font></dt><dd>
<dl>
<dt><font face="helvetica, arial"><a href="noncoding_objects.html#IgorfDetector">IgorfDetector</a>
</font></dt><dt><font face="helvetica, arial"><a href="noncoding_objects.html#Igr">Igr</a>
</font></dt></dl>
</dd>
</dl>
 <p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ffc8d8">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#000000" face="helvetica, arial"><a name="Igorf">class <strong>Igorf</strong></a>(<a href="biobjects.html#Feature">biobjects.Feature</a>)</font></td></tr>
    
<tr bgcolor="#ffc8d8"><td rowspan=2><tt>&nbsp;&nbsp;&nbsp;</tt></td>
<td colspan=2><tt>Initialize&nbsp;IGORFs&nbsp;from&nbsp;<a href="#Igorf">Igorf</a>&nbsp;detection&nbsp;results.<br>&nbsp;</tt></td></tr>
<tr><td>&nbsp;</td>
<td width="100%"><dl><dt>Method resolution order:</dt>
<dd><a href="noncoding_objects.html#Igorf">Igorf</a></dd>
<dd><a href="biobjects.html#Feature">biobjects.Feature</a></dd>
<dd><a href="biobjects.html#StraGeRa">biobjects.StraGeRa</a></dd>
<dd><a href="builtins.html#object">builtins.object</a></dd>
</dl>
<hr>
Methods defined here:<br>
<dl><dt><a name="Igorf-__init__"><strong>__init__</strong></a>(self, gffline)</dt><dd><tt>gffline&nbsp;required.</tt></dd></dl>

<dl><dt><a name="Igorf-frames_coverage"><strong>frames_coverage</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'], rm_zero=True)</dt><dd><tt>Return&nbsp;a&nbsp;dictionary&nbsp;with&nbsp;the&nbsp;coverage&nbsp;of&nbsp;the&nbsp;3&nbsp;frames.<br>
rm_zero&nbsp;is&nbsp;used&nbsp;to&nbsp;remove&nbsp;0&nbsp;in&nbsp;other&nbsp;frames&nbsp;for&nbsp;a&nbsp;given&nbsp;position.</tt></dd></dl>

<dl><dt><a name="Igorf-score_coverage"><strong>score_coverage</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'], frames=False, threshold=1)</dt><dd><tt>The&nbsp;score_coverage&nbsp;is&nbsp;defined&nbsp;by&nbsp;the&nbsp;%&nbsp;of&nbsp;positions&nbsp;with&nbsp;a&nbsp;coverage<br>
&gt;=&nbsp;threshold.&nbsp;If&nbsp;frames&nbsp;is&nbsp;True,&nbsp;this&nbsp;score&nbsp;will&nbsp;be&nbsp;computed&nbsp;for&nbsp;each<br>
frame&nbsp;separately.</tt></dd></dl>

<dl><dt><a name="Igorf-stretches"><strong>stretches</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'], frames=False, threshold=1, min_len=10)</dt><dd><tt>Return&nbsp;the&nbsp;stretch&nbsp;of&nbsp;position&nbsp;with&nbsp;coverages&nbsp;&lt;&nbsp;threshold.<br>
If&nbsp;frames&nbsp;is&nbsp;True,&nbsp;the&nbsp;stretches&nbsp;will&nbsp;be&nbsp;seeked&nbsp;in&nbsp;each&nbsp;frame&nbsp;coverage.</tt></dd></dl>

<hr>
Methods inherited from <a href="biobjects.html#Feature">biobjects.Feature</a>:<br>
<dl><dt><a name="Igorf-__repr__"><strong>__repr__</strong></a>(self)</dt><dd><tt>Convenient&nbsp;visualisation&nbsp;on&nbsp;printing</tt></dd></dl>

<dl><dt><a name="Igorf-add_tag"><strong>add_tag</strong></a>(self, tag_name, tag_value)</dt></dl>

<dl><dt><a name="Igorf-get_tag"><strong>get_tag</strong></a>(self, tag_name)</dt><dd><tt>Return&nbsp;the&nbsp;value&nbsp;of&nbsp;the&nbsp;tag_name.&nbsp;If&nbsp;not&nbsp;found,&nbsp;return&nbsp;None.</tt></dd></dl>

<dl><dt><a name="Igorf-gffline"><strong>gffline</strong></a>(self)</dt><dd><tt>Return&nbsp;the&nbsp;gff&nbsp;line&nbsp;(without&nbsp;<br>
)&nbsp;corresponding&nbsp;to&nbsp;the&nbsp;feature.</tt></dd></dl>

<dl><dt><a name="Igorf-rm_tag"><strong>rm_tag</strong></a>(self, tag_name)</dt></dl>

<dl><dt><a name="Igorf-set_tag"><strong>set_tag</strong></a>(self, tag_name, tag_value)</dt><dd><tt>Convenient&nbsp;way&nbsp;to&nbsp;set&nbsp;a&nbsp;tag&nbsp;value&nbsp;and&nbsp;check&nbsp;if&nbsp;the&nbsp;tag&nbsp;is&nbsp;legal.</tt></dd></dl>

<dl><dt><a name="Igorf-update_attributes"><strong>update_attributes</strong></a>(self)</dt><dd><tt>Update&nbsp;self.<strong>attributes</strong>&nbsp;taking&nbsp;tag&nbsp;values&nbsp;in&nbsp;self.<strong>tags</strong>.</tt></dd></dl>

<hr>
Data descriptors inherited from <a href="biobjects.html#Feature">biobjects.Feature</a>:<br>
<dl><dt><strong>ID</strong></dt>
<dd><tt>Convenient&nbsp;way&nbsp;to&nbsp;access&nbsp;ID&nbsp;tag.</tt></dd>
</dl>
<hr>
Methods inherited from <a href="biobjects.html#StraGeRa">biobjects.StraGeRa</a>:<br>
<dl><dt><a name="Igorf-__lt__"><strong>__lt__</strong></a>(self, other)</dt><dd><tt>StraGeRa&nbsp;are&nbsp;sorted&nbsp;by&nbsp;their&nbsp;start&nbsp;position&nbsp;(5').</tt></dd></dl>

<dl><dt><a name="Igorf-adjust_limits"><strong>adjust_limits</strong></a>(self, n)</dt><dd><tt>Return&nbsp;the&nbsp;adjused&nbsp;limits&nbsp;start/end&nbsp;depending&nbsp;on&nbsp;the&nbsp;number<br>
of&nbsp;positions&nbsp;required&nbsp;by&nbsp;'n'.&nbsp;The&nbsp;increment&nbsp;is&nbsp;usefull<br>
to&nbsp;conserve&nbsp;the&nbsp;orientation&nbsp;(can&nbsp;be&nbsp;1&nbsp;or&nbsp;-1).</tt></dd></dl>

<dl><dt><a name="Igorf-bokeh_figure"><strong>bokeh_figure</strong></a>(self, shift=12, side=5, fapick=None, only_total=False)</dt><dd><tt>Coverage&nbsp;representation&nbsp;of&nbsp;the&nbsp;StraGeRa.</tt></dd></dl>

<dl><dt><a name="Igorf-counts"><strong>counts</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'])</dt><dd><tt>Return&nbsp;a&nbsp;dictionary&nbsp;with&nbsp;the&nbsp;number&nbsp;of&nbsp;reads&nbsp;(each&nbsp;kmer&nbsp;detailed)<br>
on&nbsp;the&nbsp;StraGeRa&nbsp;positions.</tt></dd></dl>

<dl><dt><a name="Igorf-covered_positions"><strong>covered_positions</strong></a>(self, n=0)</dt><dd><tt>Return&nbsp;a&nbsp;generator&nbsp;of&nbsp;all&nbsp;positions&nbsp;covered&nbsp;by<br>
the&nbsp;StraGeRa&nbsp;(start&nbsp;and&nbsp;end&nbsp;are&nbsp;included).<br>
'n'&nbsp;can&nbsp;be&nbsp;used&nbsp;to&nbsp;select&nbsp;a&nbsp;sub-range.&nbsp;If&nbsp;n&nbsp;&gt;&nbsp;0,&nbsp;the<br>
positions&nbsp;are&nbsp;selected&nbsp;from&nbsp;the&nbsp;5'.&nbsp;If&nbsp;n&nbsp;&lt;&nbsp;0,&nbsp;from&nbsp;the&nbsp;3'.<br>
If&nbsp;n&nbsp;&gt;&nbsp;self.<a href="#Igorf-length">length</a>()&nbsp;or&nbsp;n&nbsp;=&nbsp;0,&nbsp;n&nbsp;is&nbsp;ignored&nbsp;and&nbsp;all&nbsp;positions<br>
are&nbsp;returned.</tt></dd></dl>

<dl><dt><a name="Igorf-length"><strong>length</strong></a>(self)</dt><dd><tt>Return&nbsp;the&nbsp;length&nbsp;of&nbsp;the&nbsp;StraGeRa.</tt></dd></dl>

<dl><dt><a name="Igorf-mask"><strong>mask</strong></a>(self, fapick)</dt><dd><tt>Mask&nbsp;the&nbsp;StraGeRa&nbsp;sequence&nbsp;in&nbsp;the&nbsp;fapick&nbsp;(fasta&nbsp;in&nbsp;pickle).<br>
/!\&nbsp;Warning,&nbsp;fapick&nbsp;is&nbsp;modified&nbsp;inplace!</tt></dd></dl>

<dl><dt><a name="Igorf-pos_cov"><strong>pos_cov</strong></a>(self, bamfile, shift=12, side=5, kmers=['Total'], n=0)</dt><dd><tt>Return&nbsp;a&nbsp;dictionary&nbsp;with&nbsp;the&nbsp;coverage&nbsp;informations&nbsp;with&nbsp;the&nbsp;keys:<br>
-&nbsp;'positions'&nbsp;=&nbsp;all&nbsp;the&nbsp;genomic&nbsp;positions&nbsp;covered&nbsp;by&nbsp;StraGeRa&nbsp;(5'=&gt;3')<br>
-&nbsp;'coverages'&nbsp;=&nbsp;#&nbsp;of&nbsp;reads&nbsp;on&nbsp;each&nbsp;positions,&nbsp;same&nbsp;order&nbsp;as&nbsp;'positions'<br>
-&nbsp;'details'&nbsp;=&nbsp;list&nbsp;of&nbsp;dictionaries&nbsp;with&nbsp;the&nbsp;Kmer&nbsp;count&nbsp;details<br>
-&nbsp;'bank_size'&nbsp;=&nbsp;total&nbsp;number&nbsp;of&nbsp;read&nbsp;mapped&nbsp;in&nbsp;the&nbsp;bamfile</tt></dd></dl>

<dl><dt><a name="Igorf-sequence"><strong>sequence</strong></a>(self, fapick, n=0, translate_to_protein=False)</dt><dd><tt>Return&nbsp;the&nbsp;5'&nbsp;-&gt;&nbsp;3'&nbsp;oriented&nbsp;sequence&nbsp;of&nbsp;the&nbsp;StraGeRa.<br>
`fapick`&nbsp;=&nbsp;&lt;dict&gt;&nbsp;with&nbsp;seqid&nbsp;as&nbsp;key&nbsp;and&nbsp;sequence&nbsp;as&nbsp;value.<br>
Use&nbsp;'fasta2pickle.py'&nbsp;program&nbsp;to&nbsp;generate&nbsp;a&nbsp;fapick&nbsp;from&nbsp;Fasta.</tt></dd></dl>

<dl><dt><a name="Igorf-tupleview"><strong>tupleview</strong></a>(self, strand='int')</dt><dd><tt>Return&nbsp;a&nbsp;tuple&nbsp;view&nbsp;of&nbsp;the&nbsp;StraGeRa&nbsp;with:<br>
(start,&nbsp;end,&nbsp;increment/strand)<br>
MIN_COORDS&nbsp;is&nbsp;a&nbsp;namedtuple&nbsp;defined&nbsp;in&nbsp;constants.py.<br>
&nbsp;<br>
I&nbsp;think&nbsp;I&nbsp;need&nbsp;to&nbsp;remove&nbsp;it,&nbsp;and&nbsp;use&nbsp;StraGeRa&nbsp;objects&nbsp;instead?</tt></dd></dl>

<hr>
Data descriptors inherited from <a href="biobjects.html#StraGeRa">biobjects.StraGeRa</a>:<br>
<dl><dt><strong>__dict__</strong></dt>
<dd><tt>dictionary&nbsp;for&nbsp;instance&nbsp;variables&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
<dl><dt><strong>__weakref__</strong></dt>
<dd><tt>list&nbsp;of&nbsp;weak&nbsp;references&nbsp;to&nbsp;the&nbsp;object&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
</td></tr></table> <p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ffc8d8">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#000000" face="helvetica, arial"><a name="IgorfDetector">class <strong>IgorfDetector</strong></a>(<a href="builtins.html#object">builtins.object</a>)</font></td></tr>
    
<tr bgcolor="#ffc8d8"><td rowspan=2><tt>&nbsp;&nbsp;&nbsp;</tt></td>
<td colspan=2><tt>Store&nbsp;IGORFs&nbsp;attributes&nbsp;for&nbsp;the&nbsp;detection.<br>&nbsp;</tt></td></tr>
<tr><td>&nbsp;</td>
<td width="100%">Methods defined here:<br>
<dl><dt><a name="IgorfDetector-__init__"><strong>__init__</strong></a>(self, igr_parent, start, stop, strand, increment, sequence, absolute_frame)</dt><dd><tt>Init&nbsp;<a href="#Igorf">Igorf</a>&nbsp;as&nbsp;'strict'&nbsp;and&nbsp;'IG'</tt></dd></dl>

<dl><dt><a name="IgorfDetector-gffline"><strong>gffline</strong></a>(self)</dt><dd><tt>Return&nbsp;the&nbsp;GFF&nbsp;line&nbsp;correponding&nbsp;to&nbsp;the&nbsp;igorf.</tt></dd></dl>

<dl><dt><a name="IgorfDetector-update_left"><strong>update_left</strong></a> = update(self, *args, **kwargs)</dt></dl>

<dl><dt><a name="IgorfDetector-update_right"><strong>update_right</strong></a> = update(self, *args, **kwargs)</dt></dl>

<hr>
Data descriptors defined here:<br>
<dl><dt><strong>__dict__</strong></dt>
<dd><tt>dictionary&nbsp;for&nbsp;instance&nbsp;variables&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
<dl><dt><strong>__weakref__</strong></dt>
<dd><tt>list&nbsp;of&nbsp;weak&nbsp;references&nbsp;to&nbsp;the&nbsp;object&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
</td></tr></table> <p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ffc8d8">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#000000" face="helvetica, arial"><a name="Igr">class <strong>Igr</strong></a>(<a href="builtins.html#object">builtins.object</a>)</font></td></tr>
    
<tr bgcolor="#ffc8d8"><td rowspan=2><tt>&nbsp;&nbsp;&nbsp;</tt></td>
<td colspan=2><tt>IGR&nbsp;strands&nbsp;for&nbsp;InterGenic&nbsp;Region.<br>&nbsp;</tt></td></tr>
<tr><td>&nbsp;</td>
<td width="100%">Methods defined here:<br>
<dl><dt><a name="Igr-__init__"><strong>__init__</strong></a>(self, seqid, left_genomic, right_genomic, left_strand='', right_strand='', left_frame=None, right_frame=None)</dt><dd><tt>constructor&nbsp;v1:&nbsp;strands&nbsp;are&nbsp;-1&nbsp;('-')&nbsp;and&nbsp;+1&nbsp;('+')</tt></dd></dl>

<dl><dt><a name="Igr-__lt__"><strong>__lt__</strong></a>(self, other)</dt><dd><tt>sorted(igr_list)&nbsp;will&nbsp;sort&nbsp;igrs&nbsp;on&nbsp;their&nbsp;left_genomic.</tt></dd></dl>

<dl><dt><a name="Igr-get_sequence"><strong>get_sequence</strong></a>(self, fapick)</dt><dd><tt>Get&nbsp;the&nbsp;IGR&nbsp;sequence,&nbsp;strand&nbsp;+,&nbsp;from&nbsp;the&nbsp;fasta&nbsp;pickle.</tt></dd></dl>

<dl><dt><a name="Igr-gffline"><strong>gffline</strong></a>(self)</dt><dd><tt>Return&nbsp;the&nbsp;gffline&nbsp;corresponding&nbsp;to&nbsp;the&nbsp;IGR</tt></dd></dl>

<dl><dt><a name="Igr-tuple_view"><strong>tuple_view</strong></a>(self)</dt><dd><tt>Return&nbsp;a&nbsp;tupleview&nbsp;of&nbsp;the&nbsp;IGR.</tt></dd></dl>

<hr>
Data descriptors defined here:<br>
<dl><dt><strong>__dict__</strong></dt>
<dd><tt>dictionary&nbsp;for&nbsp;instance&nbsp;variables&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
<dl><dt><strong>__weakref__</strong></dt>
<dd><tt>list&nbsp;of&nbsp;weak&nbsp;references&nbsp;to&nbsp;the&nbsp;object&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
</td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#55aa55">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Data</strong></big></font></td></tr>
    
<tr><td bgcolor="#55aa55"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><strong>APPRIS_LEVEL</strong> = 1<br>
<strong>APPRIS_P_ALL</strong> = 'appris_principal'<br>
<strong>APPRIS_P_LEV</strong> = 'appris_principal_1'<br>
<strong>ATTRIBUTES</strong> = 8<br>
<strong>BLIT</strong> = '<font color="#c040c0">\x1b</font>[3;34m'<br>
<strong>BLITUN</strong> = '<font color="#c040c0">\x1b</font>[4;34m'<br>
<strong>CDS</strong> = 'CDS'<br>
<strong>END</strong> = 4<br>
<strong>EXON</strong> = 'exon'<br>
<strong>FRAMES_CORREL</strong> = {0: {0: 0, 1: 1, 2: 2}, 1: {0: 2, 1: 0, 2: 1}, 2: {0: 1, 1: 2, 2: 0}}<br>
<strong>GENE</strong> = 'gene'<br>
<strong>GFF_COM</strong> = '#'<br>
<strong>GFF_SEP</strong> = '<font color="#c040c0">\t</font>'<br>
<strong>GRIT</strong> = '<font color="#c040c0">\x1b</font>[3;32m'<br>
<strong>ID_TAG</strong> = 'ID'<br>
<strong>IG_COLS</strong> = {'+': {0: '#ff4d4d', 1: '#ff4d4d', 2: '#ff4d4d'}, '-': {0: '#3366ff', 1: '#3366ff', 2: '#3366ff'}, 'IGR': '#2eb82e'}<br>
<strong>INNER</strong> = 100<br>
<strong>KMERS</strong> = ['Total']<br>
<strong>MASK_BASE</strong> = 'N'<br>
<strong>NA</strong> = '<font color="#c040c0">\x1b</font>[0m'<br>
<strong>NAME_TAG</strong> = 'Name'<br>
<strong>N_0</strong> = 0<br>
<strong>ORDERED_REGIONS</strong> = ['UTR5', 'CDS', 'UTR3']<br>
<strong>ORIT</strong> = '<font color="#c040c0">\x1b</font>[3;33m'<br>
<strong>OUTER</strong> = 20<br>
<strong>PARENT_TAG</strong> = 'Parent'<br>
<strong>PHASE</strong> = 7<br>
<strong>RANK_TAG</strong> = 'exon_number'<br>
<strong>REIT</strong> = '<font color="#c040c0">\x1b</font>[3;31m'<br>
<strong>SCORE</strong> = 5<br>
<strong>SEQID</strong> = 0<br>
<strong>SHIFT</strong> = 12<br>
<strong>SIDE</strong> = 5<br>
<strong>SOURCE</strong> = 1<br>
<strong>START</strong> = 3<br>
<strong>STOPS</strong> = ['TAG', 'TAA', 'TGA']<br>
<strong>STOPS_REV</strong> = ['TTA', 'CTA', 'TCA']<br>
<strong>STRAND</strong> = 6<br>
<strong>STRANDS_MAP</strong> = {'+': {'ali': 0, 'ali_rev': 16, 'one': 1}, '-': {'ali': 16, 'ali_rev': 0, 'one': -1}, '.': {'one': 1}}<br>
<strong>TAG_SEP</strong> = ';'<br>
<strong>TOTAL_KEYNAME</strong> = 'Total'<br>
<strong>TRANSCRIPT</strong> = 'transcript'<br>
<strong>TYPE</strong> = 2<br>
<strong>UTR</strong> = 'UTR'<br>
<strong>UTR3</strong> = 'UTR3'<br>
<strong>UTR5</strong> = 'UTR5'<br>
<strong>VAL_SEP</strong> = '='<br>
<strong>__email__</strong> = 'pierre.bertin@i2bc.paris-saclay.fr'<br>
<strong>__license__</strong> = 'GPL'<br>
<strong>__maintainer__</strong> = 'Pierre Bertin'<br>
<strong>__status__</strong> = 'Development'</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#7799ee">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Author</strong></big></font></td></tr>
    
<tr><td bgcolor="#7799ee"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%">Pierre&nbsp;Bertin</td></tr></table>
</body></html>