<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module sambam</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong>sambam</strong></big></big> (version 1.0.1)</font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="file:/media/jean-noel/98257acf-f310-4578-90bb-4121919162c6/Projet_Anne/strali/sambam.py">/media/jean-noel/98257acf-f310-4578-90bb-4121919162c6/Projet_Anne/strali/sambam.py</a></font></td></tr></table>
    <p><tt>Module&nbsp;to&nbsp;handle&nbsp;SAM/BAM&nbsp;files.<br>
&nbsp;<br>
The&nbsp;notion&nbsp;of&nbsp;'alignment&nbsp;strand'&nbsp;is&nbsp;introduced&nbsp;here.<br>
For&nbsp;RIBO-SEQ&nbsp;data,&nbsp;the&nbsp;flags&nbsp;corresponding&nbsp;to&nbsp;the<br>
alignment&nbsp;strands&nbsp;are&nbsp;0&nbsp;('+')&nbsp;and&nbsp;16&nbsp;('-').</tt></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#aa55cc">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Modules</strong></big></font></td></tr>
    
<tr><td bgcolor="#aa55cc"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><table width="100%" summary="list"><tr><td width="25%" valign=top><a href="os.html">os</a><br>
</td><td width="25%" valign=top><a href="pysam.html">pysam</a><br>
</td><td width="25%" valign=top></td><td width="25%" valign=top></td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#eeaa77">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Functions</strong></big></font></td></tr>
    
<tr><td bgcolor="#eeaa77"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl><dt><a name="-count_all_kmers"><strong>count_all_kmers</strong></a>(chromosome, ali_strand, start, end, bamfile, reduced_index)</dt><dd><tt>Populate&nbsp;all_kmers_counts&nbsp;with&nbsp;all&nbsp;the&nbsp;Kmers&nbsp;found&nbsp;in&nbsp;the&nbsp;bamfile.</tt></dd></dl>
 <dl><dt><a name="-count_selected_kmers"><strong>count_selected_kmers</strong></a>(chromosome, ali_strand, start, end, bamfile, reduced_index, kmers)</dt><dd><tt>Populate&nbsp;the&nbsp;kmers_counts&nbsp;with&nbsp;the&nbsp;selected&nbsp;kmers&nbsp;and&nbsp;the&nbsp;total.</tt></dd></dl>
 <dl><dt><a name="-get_bank_size"><strong>get_bank_size</strong></a>(bamfile)</dt><dd><tt>Return&nbsp;the&nbsp;total&nbsp;number&nbsp;of&nbsp;reads&nbsp;mapped&nbsp;in&nbsp;the&nbsp;bamfile.</tt></dd></dl>
 <dl><dt><a name="-get_env_reduced_pos"><strong>get_env_reduced_pos</strong></a>(reference, strand, five_prime_1b, three_prime_1b, bamfile, meta_codon, ref_pos_1b, shift=12, side=5, kmers=['Total'])</dt><dd><tt>Automatically&nbsp;fill&nbsp;the&nbsp;meta_codon&nbsp;with&nbsp;the&nbsp;relative&nbsp;position&nbsp;between<br>
five_prime1b&nbsp;and&nbsp;three_prime_1b.</tt></dd></dl>
 <dl><dt><a name="-get_shift_idx"><strong>get_shift_idx</strong></a>(shift, side, ali_strand)</dt><dd><tt>Return&nbsp;the&nbsp;reduced&nbsp;position&nbsp;index&nbsp;with&nbsp;shift&nbsp;applied.<br>
shift&nbsp;=&nbsp;0&nbsp;means&nbsp;5'&nbsp;or&nbsp;3'&nbsp;most&nbsp;positions.&nbsp;Side&nbsp;can&nbsp;be&nbsp;3&nbsp;or&nbsp;5,<br>
and&nbsp;ali_strand&nbsp;0&nbsp;or&nbsp;16.</tt></dd></dl>
 <dl><dt><a name="-get_stranded_region_counts"><strong>get_stranded_region_counts</strong></a>(chromosome, strand, start_1_based, end_1_based, bamfile, shift=12, side=5, kmers=['Total'])</dt><dd><tt>Return&nbsp;a&nbsp;dictionary&nbsp;with&nbsp;the&nbsp;count&nbsp;of&nbsp;each&nbsp;kmers<br>
between&nbsp;start&nbsp;and&nbsp;end.&nbsp;Only&nbsp;stranded&nbsp;reads&nbsp;are&nbsp;kept.</tt></dd></dl>
 <dl><dt><a name="-get_stranded_region_coverage_reduced"><strong>get_stranded_region_coverage_reduced</strong></a>(chromosome, strand, start_1_based, end_1_based, bamfile, shift=12, side=5, kmers=['Total'])</dt><dd><tt>Return&nbsp;a&nbsp;dictionary&nbsp;with&nbsp;the&nbsp;coverage&nbsp;of&nbsp;each&nbsp;position&nbsp;in&nbsp;the&nbsp;given<br>
region.&nbsp;The&nbsp;reads&nbsp;are&nbsp;reduced&nbsp;to&nbsp;only&nbsp;one&nbsp;position,&nbsp;defined&nbsp;by&nbsp;the&nbsp;'shift'<br>
and&nbsp;the&nbsp;'side'.&nbsp;'kmers'&nbsp;is&nbsp;used&nbsp;to&nbsp;selected&nbsp;specific&nbsp;sizes&nbsp;of&nbsp;reads.</tt></dd></dl>
 <dl><dt><a name="-init_pysam_bam"><strong>init_pysam_bam</strong></a>(bamfile_path, auto_sort=False)</dt><dd><tt>Check&nbsp;if&nbsp;the&nbsp;bamfile&nbsp;has&nbsp;an&nbsp;index.&nbsp;If&nbsp;not,&nbsp;the&nbsp;file&nbsp;is&nbsp;indexed.<br>
If&nbsp;it&nbsp;can&nbsp;be&nbsp;indexed,&nbsp;it&nbsp;will&nbsp;be&nbsp;sorted.<br>
Returns&nbsp;a&nbsp;pysam.AlignmentFile.</tt></dd></dl>
 <dl><dt><a name="-select_kmers"><strong>select_kmers</strong></a>(kmers, pos_cov)</dt><dd><tt>Select&nbsp;the&nbsp;kmers&nbsp;in&nbsp;`kmers`&nbsp;for&nbsp;each&nbsp;positions<br>
in&nbsp;pos_cov.&nbsp;Return&nbsp;a&nbsp;new&nbsp;dictionary&nbsp;with&nbsp;TOTAL&nbsp;adjusted.</tt></dd></dl>
</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#55aa55">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Data</strong></big></font></td></tr>
    
<tr><td bgcolor="#55aa55"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><strong>APPRIS_LEVEL</strong> = 1<br>
<strong>APPRIS_P_ALL</strong> = 'appris_principal'<br>
<strong>APPRIS_P_LEV</strong> = 'appris_principal_1'<br>
<strong>ATTRIBUTES</strong> = 8<br>
<strong>BLIT</strong> = '<font color="#c040c0">\x1b</font>[3;34m'<br>
<strong>BLITUN</strong> = '<font color="#c040c0">\x1b</font>[4;34m'<br>
<strong>CDS</strong> = 'CDS'<br>
<strong>END</strong> = 4<br>
<strong>EXON</strong> = 'exon'<br>
<strong>FRAMES_CORREL</strong> = {0: {0: 0, 1: 1, 2: 2}, 1: {0: 2, 1: 0, 2: 1}, 2: {0: 1, 1: 2, 2: 0}}<br>
<strong>GENE</strong> = 'gene'<br>
<strong>GFF_COM</strong> = '#'<br>
<strong>GFF_SEP</strong> = '<font color="#c040c0">\t</font>'<br>
<strong>GRIT</strong> = '<font color="#c040c0">\x1b</font>[3;32m'<br>
<strong>ID_TAG</strong> = 'ID'<br>
<strong>INNER</strong> = 100<br>
<strong>KMERS</strong> = ['Total']<br>
<strong>MASK_BASE</strong> = 'N'<br>
<strong>NA</strong> = '<font color="#c040c0">\x1b</font>[0m'<br>
<strong>NAME_TAG</strong> = 'Name'<br>
<strong>N_0</strong> = 0<br>
<strong>ORDERED_REGIONS</strong> = ['UTR5', 'CDS', 'UTR3']<br>
<strong>ORIT</strong> = '<font color="#c040c0">\x1b</font>[3;33m'<br>
<strong>OUTER</strong> = 20<br>
<strong>PARENT_TAG</strong> = 'Parent'<br>
<strong>PHASE</strong> = 7<br>
<strong>RANK_TAG</strong> = 'exon_number'<br>
<strong>REIT</strong> = '<font color="#c040c0">\x1b</font>[3;31m'<br>
<strong>SCORE</strong> = 5<br>
<strong>SEQID</strong> = 0<br>
<strong>SHIFT</strong> = 12<br>
<strong>SIDE</strong> = 5<br>
<strong>SOURCE</strong> = 1<br>
<strong>START</strong> = 3<br>
<strong>STRAND</strong> = 6<br>
<strong>STRANDS_MAP</strong> = {'+': {'ali': 0, 'ali_rev': 16, 'one': 1}, '-': {'ali': 16, 'ali_rev': 0, 'one': -1}, '.': {'one': 1}}<br>
<strong>TAG_SEP</strong> = ';'<br>
<strong>TOTAL_KEYNAME</strong> = 'Total'<br>
<strong>TRANSCRIPT</strong> = 'transcript'<br>
<strong>TYPE</strong> = 2<br>
<strong>UTR</strong> = 'UTR'<br>
<strong>UTR3</strong> = 'UTR3'<br>
<strong>UTR5</strong> = 'UTR5'<br>
<strong>VAL_SEP</strong> = '='<br>
<strong>__email__</strong> = 'pierre.bertin@i2bc.paris-saclay.fr'<br>
<strong>__license__</strong> = 'GPL'<br>
<strong>__maintainer__</strong> = 'Pierre Bertin'<br>
<strong>__status__</strong> = 'Development'</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#7799ee">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Author</strong></big></font></td></tr>
    
<tr><td bgcolor="#7799ee"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%">Pierre&nbsp;Bertin</td></tr></table>
</body></html>